README.md
# mbdns

`mbdns` is a dynamic DNS update client for the [Mythic Beasts](https://www.mythic-beasts.com/support/api/dnsv2) Primary DNS v2 system, supporting their IPv4 and IPv6 endpoints. Written in [golang](https://golang.org), it supports common home network infrastructure operating systems (FreeBSD, EdgeOS, UniFi OS, Linux) and common platform architectures (amd64, arm64 and MIPS).

## Configuring mbdns

Take [doc/mbdns.conf.sample](/doc/mbdns.conf.sample) and configure it to your needs, saving as `mbdns.conf`. `mbdns.conf` must be valid JSON or `mbdns` will fail to start.

## Deploying mbdns

* Copy the `mbdns` binary (usually named `mbdns-VERSION-OS-ARCH`) and `mbdns.conf` to your target platform
* chmod 0400 mbdns.conf so the Mythic Beasts API secrets can only be read by the user you deploy as (ideally use a specific user). `mbdns` will check for you and fail to start if the config is insecurely readable, unless you use `--insecure`.
* Run the `mbdns` binary, which will process its main update until you kill it
* By default the `mbdns` binary will look for `./mbdns.conf`. You can relocate it (and rename it) and tell `mbdns` with `--config`.
  * `mbdns --config /etc/mbdns/mbdns.conf`

`mbdns` is written in golang and builds as a statically linked library with no dependencies.

## Practical running

`mbdns` logs to `stdout`. If your target OS supports we suggest you redirect stdout to a file, logrotate that file, and run the binary in the background via the daemonising system of your choice. `mbdns` makes no attempt to daemonise itself.

[systemd](https://systemd.io/) will run it nicely if you use the `simple` service type, running the process for you and capturing `stdout`.

`mbdns --version` prints the version information and exits immediately.

## Logging

`mbdns` logs the following:

* version, build date and git commit SHA on startup
* the path to the config file
* success or failure including record type
* a small handful of at-startup failure messages if it can't run (no config, invalid JSON in the config, insecure config)
* a message saying it is processing records if it can cleanly start

## Platform specific instructions

- [UniFi Dream Machine SE](/doc/udm-se/README.md)

## License

`mbdns` is MIT licensed.

CHANGELOG
# 08-09-2018

Functioning implementation tested on `darwin-amd64`, `freebsd-amd64` and `linux-mipsle`

+ [x] `darwin-amd64` (macOS 10.13)
+ [x] `freebsd-amd64` (FreeNAS 11)
+ [x] `linux-mipsle` (EdgeOS 1.10, ER-X)
+ [x] Runs from secured JSON-based config
+ [x] Supports `A` and `AAAA` record types against the correct API endpoint
+ [x] Error diagnosis on non-200 returns
+ [x] 1s delay between record update attempts (compile time)
+ [x] 300s delay between full loop update attempts (compile time)
+ [x] Scripts for building a release with embedded `BuildVersion`, `BuildDate` and `GitRev` (printed to console at start)

# 09-09-2018

+ [x] Support relocating and renaming `./mbdns.conf` with `--config`

# 23-01-2019

+ [x] Fixed a bug where the API response body would read erroneously on error

# 11-05-2022

+ [x] [docs for the UDM-SE](/doc/udm-se/README.md)
+ [x] added support for linux arm64/aarch64 builds for the UDM-SE
+ [x] rebuilt all `release-1.0.2` binaries with go1.18.1

# 15-05-2022

+ [x] Support for the [Mythic Beasts Primary DNS API v2](https://www.mythic-beasts.com/support/api/dnsv2)
+ [x] Proper go module support
+ [x] Reworked documentation to suit the new implementation

mbdns.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

// Simple type to help with record type disambiguation
type MbdnsRecordType string

const (
	mbRecordTypeA    MbdnsRecordType = "A"
	mbRecordTypeAAAA MbdnsRecordType = "AAAA"
)

// Type to help us unmarshal the JSON config from disk
type MbdnsConfig struct {
	Zone   string          `json:"zone"`
	Host   string          `json:"host"`
	KeyID  string          `json:"key_id"`
	Secret string          `json:"secret"`
	Type   MbdnsRecordType `json:"type"`
}

// Main error type and strings used when calling the HTTPS API endpoints
type MbdnsUpdateRecordError struct {
	Err error
}

func (m *MbdnsUpdateRecordError) Error() string {
	return fmt.Sprintf("update record error: %v", m.Err)
}

const (
	mbErrorClientConnection string = "ERROR_CLIENT_CONNECTION"
	mbErrorClientResponse   string = "ERROR_CLIENT_RESPONSE"
	mbErrorRequestCreation  string = "ERROR_REQUEST_CREATION"
	mbErrorUpdateFailure    string = "ERROR_UPDATE_FAILURE"
)

// Log strings printed when starting up
const (
	mbLogNoConfig            string = "config file does not exist"
	mbLogFailedConfigStat    string = "failed to stat config file"
	mbLogInsecureConfig      string = "config file is insecure"
	mbLogReadingConfig       string = "mbdns is reading its config"
	mbLogCannotReadConfig    string = "cannot read config file"
	mbLogCannotProcessConfig string = "cannot process config file"
	mbLogProcessingRecords   string = "mbdns is processing records"
)

// Help texts
const (
	mbHelpConfigFilePath      string = "Config file path"
	mbHelpPrintVersion        string = "Print version banner and exit"
	mbHelpAllowInsecureConfig string = "Allow reading an insecure config"
)

// Main application configuration and logging
const (
	mbConfigFile              string = "mbdns.conf"
	mbURLv4_2                 string = "https://ipv4.api.mythic-beasts.com/dns/v2/dynamic/%s.%s"
	mbURLv6_2                 string = "https://ipv6.api.mythic-beasts.com/dns/v2/dynamic/%s.%s"
	mbLoopWaitSeconds         string = "300s"
	mbRecordUpdateWaitSeconds string = "1s"
	mbStartupBanner           string = "mbdns %s %s (git %s) by %s"
	mbConfigPathBanner        string = "config path: %s"
	mbUpdateResponseError     string = "updating %s.%s (%s) failed with %s"
	mbUpdateResponseSuccess   string = "updating %s.%s (%s) succeeded"
)

// unmarshalled running configuration
var mbdnsConfig []MbdnsConfig

// Build-time version information
var BuildVersion string
var BuildDate string
var GitRev string
var BuildUser string

// call the Mythic Beasts DNS v2 API endpoint for dynamic DNS
func updateDynamicDNSv2(url string, clientID string, clientSecret string) error {
	request, err := http.NewRequest(http.MethodPost, url, strings.NewReader(""))

	if err != nil {
		return &MbdnsUpdateRecordError{Err: errors.New(mbErrorRequestCreation)}
	}

	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	request.SetBasicAuth(clientID, clientSecret)

	client := &http.Client{}

	response, err := client.Do(request)

	if err != nil {
		return &MbdnsUpdateRecordError{Err: errors.New(mbErrorUpdateFailure)}
	}

	if response.StatusCode != 200 {
		return &MbdnsUpdateRecordError{Err: errors.New(mbErrorClientResponse + " code: " + fmt.Sprint(response.StatusCode))}
	}

	return nil
}

// main processing loop that works on config records
func processRecords() {
	loopSleepDuration, _ := time.ParseDuration(mbLoopWaitSeconds)
	recordSleepDuration, _ := time.ParseDuration(mbRecordUpdateWaitSeconds)

	for {
		for i := range mbdnsConfig {
			mbURL := fmt.Sprintf(mbURLv4_2, mbdnsConfig[i].Host, mbdnsConfig[i].Zone)

			if mbdnsConfig[i].Type == mbRecordTypeAAAA {
				mbURL = fmt.Sprintf(mbURLv6_2, mbdnsConfig[i].Host, mbdnsConfig[i].Zone)
			}

			err := updateDynamicDNSv2(mbURL, mbdnsConfig[i].KeyID, mbdnsConfig[i].Secret)

			if err != nil {
				log.Printf(mbUpdateResponseError, mbdnsConfig[i].Host, mbdnsConfig[i].Zone, mbdnsConfig[i].Type, err.Error())
				continue
			}

			log.Printf(mbUpdateResponseSuccess, mbdnsConfig[i].Host, mbdnsConfig[i].Zone, mbdnsConfig[i].Type)

			time.Sleep(recordSleepDuration)
		}

		time.Sleep(loopSleepDuration)
	}
}

func main() {
	log.SetOutput(os.Stdout)

	var configFile string
	var printVer bool
	var insecureConf bool

	flag.StringVar(&configFile, "config", mbConfigFile, mbHelpConfigFilePath)
	flag.BoolVar(&printVer, "version", false, mbHelpPrintVersion)
	flag.BoolVar(&insecureConf, "insecure", false, mbHelpAllowInsecureConfig)
	flag.Parse()

	if printVer {
		fmt.Printf(mbStartupBanner, BuildVersion, BuildDate, GitRev, BuildUser)
		os.Exit(0)
	}

	log.Printf(mbStartupBanner, BuildVersion, BuildDate, GitRev, BuildUser)
	log.Printf(mbConfigPathBanner, configFile)

	if _, err := os.Stat(configFile); os.IsNotExist(err) {
		log.Fatal(mbLogNoConfig)
	}

	f, err := os.Lstat(configFile)

	if err != nil {
		log.Fatal(mbLogFailedConfigStat)
	}

	if f.Mode() != 0400 {
		if !insecureConf {
			log.Fatal(mbLogInsecureConfig)
		}
	}

	log.Println(mbLogReadingConfig)

	tuples, err := ioutil.ReadFile(configFile)

	if err != nil {
		log.Fatal(mbLogCannotReadConfig)
	}

	err = json.Unmarshal(tuples, &mbdnsConfig)

	if err != nil {
		log.Fatal(mbLogCannotProcessConfig)
	}

	log.Println(mbLogProcessingRecords)

	go processRecords()
	select {}
}
